libs/core/tests/unit_tests/test_tools.py

```diff
@@ -2110,3 +2110,132 @@ def injected_tool(x: int, foo: Annotated[Foo, InjectedToolArg]) -> str:
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @tool
+        def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+    assert a.foo.invoke({"a": 1, "b": 2}) == 13
+
+
+def test_method_tool_args() -> None:
+    """Test that a method tool's args do not include self."""
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @tool
+        def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+    assert "self" not in a.foo.args
+
+
+async def test_method_tool_async() -> None:
+    """Test that a method tool can be async."""
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @tool
+        async def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+    async_response = await a.foo.ainvoke({"a": 1, "b": 2})
+    assert async_response == 13
+
+
+def test_method_tool_string_invoke() -> None:
+    """Test that a method tool can be invoked with a string."""
+
+    class A:
+        def __init__(self, a: str):
+            self.a = a
+
+        @tool
+        def foo(self, b: str) -> str:
+            """Concatenate a and b."""
+            return self.a + b
+
+    a = A("a")
+    assert a.foo.invoke("b") == "ab"
+
+
+def test_method_tool_toolcall_invoke() -> None:
+    """Test that a method tool can be invoked with a ToolCall."""
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @tool
+        def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+
+    tool_call = {
+        "name": a.foo.name,
+        "args": {"a": 1, "b": 2},
+        "id": "123",
+        "type": "tool_call",
+    }
+
+    tool_message = a.foo.invoke(tool_call)
+
+    assert int(tool_message.content) == 13
+
+
+@pytest.mark.skipif(
+    sys.version_info >= (3, 13),
+    reason="Class methods wrapping properties was removed with Python 3.13",
+)
+def test_method_tool_classmethod() -> None:
+    """Test that a method tool can be a classmethod."""
+
+    class A:
+        c = 5
+
+        @classmethod
+        @tool
+        def foo(cls, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + cls.c
+
+    assert A.foo.invoke({"a": 1, "b": 2}) == 8
+
+
+@pytest.mark.skipif(
+    sys.version_info >= (3, 13),
+    reason="Class methods wrapping properties was removed with Python 3.13",
+)
+def test_method_tool_classmethod_args() -> None:
+    """Test that a classmethod tool's args do not include cls."""
+
+    class A:
+        c = 5
+
+        @classmethod
+        @tool
+        def foo(cls, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + cls.c
```

Comment:
this case actually doesn't matter - I think the one that matters is this
```suggestion
        @tool
        @classmethod
        def foo(cls, a: int, b: int) -> int:
            """Add two numbers to c."""
            return a + b + cls.c
```

Reply:
from all these differences, what are your thoughts on introducing a new decorator `tool_property` or something to distinguish these cases? Would we also need a `tool_classmethod` to be separate too?

Reply:
Some additional thoughts:
- in principle, it makes sense to force the user to distinguish rather than try to auto-detect that your thing is a property/classmethod
- especially for crazy edge cases like below, where the tool decorator would run in a non-class context, and then screw up later

```
@tool
def my_tool(...):
    ...

class A():
    it = my_tool

A.it2 = my_tool

a = A()
a.it3 = my_tool
```

Reply:
Having a separate decorator specifically for implementing this feature was always an idea, and is definitely the better solution.

To be honest, the reason things are implemented the way they currently are is because I thought that if I did not implement this in a way that satisfied the example code given by Eugene in the original issue then that issue would not be resolved.

Now that I know there is more leeway and freedom to implement this feature in the way that makes most sense rather than being locked into using the original tool decorator, that can absolutely be done.

How about I first create some sort of new decorator that is for method tools, and then from there I can see if classmethods are doable?

Reply:
doing it as 2 separate PRs is also good practice. If you'd like to avoid double-work, you're welcome to put up a PR with just the unit tests first, just to make sure we're envisioning the same interface

Reply:
Check out my new PR https://github.com/langchain-ai/langchain/pull/28695.

On top of doing what you suggested, I also came up with a much more elegant solution without any strange workarounds or shortcomings. The actual implementation is only something like 20 lines of added logic. Taking a second shot at this was definitely worth it.